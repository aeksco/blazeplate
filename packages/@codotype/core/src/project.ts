import { SchemaInput } from "./schema";
import { ConfigurationValue } from "./configuration-value";
import { TokenCasing } from "./token";
import { Schema } from "./";
import { UUID } from "./uuid";
import { RelationInput } from "./relation";

// // // //

/**
 * ProjectInput
 * Input describing a single Project for a Plugin
 * Is later transformed into Project interface when building Plugin output in Runtime
 */
export interface ProjectInput {
    id: UUID;
    pluginID: string;
    pluginVersion: string;
    identifiers: TokenCasing;
    configuration: ConfigurationValue;
    schemas: SchemaInput[];
    relations: RelationInput[];
}

/**
 * Project
 * The verified and normalized ProjectInput, optimized for use by Plugins for building output
 */
export interface Project {
    id: UUID;
    schemas: Schema[];
    identifiers: TokenCasing;
    pluginID: string;
    pluginVersion: string;
    configuration: ConfigurationValue;
}
