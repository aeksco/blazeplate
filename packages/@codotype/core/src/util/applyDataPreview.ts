import {
    DataPreview,
    DataPreviewActionTypes,
    DataPreviewConstraintTypes,
    DataPreviewRule,
    DataPreviewConstraint,
    DataPreviewAction,
} from "../property-preview";
import { OptionValueInstance } from "../configuration-property";

// // // //

/**
 * shouldApplyDataPreviewRule
 * Returns a boolean indicating whether or not a specific DataPreviewRule should be applied against props.data
 * @param props.data - The data that's being evaluated by the DataPreviewRule
 * @param props.rule - The DataPreviewRule being evaluated
 * TODO - write tests for this
 * TODO - write tests for this
 */
export function shouldApplyDataPreviewRule(props: {
    data: Record<string, OptionValueInstance>;
    constraint: DataPreviewConstraint;
}): boolean {
    const { constraint, data } = props;

    const sourceValue: any = data[constraint.dataProperty];

    // Handle EQUALS
    if (constraint.type === DataPreviewConstraintTypes.equals) {
        if (sourceValue === constraint.value) {
            return true;
        }
    }

    // Handle contains
    if (
        constraint.type === DataPreviewConstraintTypes.contains &&
        typeof sourceValue === "string"
    ) {
        return sourceValue.includes(constraint.value);
    }

    // Handle exists
    if (constraint.type === DataPreviewConstraintTypes.exists) {
        if (sourceValue) {
            return true;
        }
    }

    // Return false by default
    return false;
}

/**
 * applyDataPreviewRule
 * @param props.data - The data that's being evaluated by the DataPreviewRule
 * @param props.rule - The DataPreviewRule being evaluated
 */
export function applyDataPreviewRule(props: {
    data: Record<string, OptionValueInstance>;
    action: DataPreviewAction;
}): string {
    const { action, data } = props;

    // Handle LITERAL
    if (action.type === DataPreviewActionTypes.literal) {
        return action.template;
    }

    // Handle STRING_TEMPLATE
    if (action.type === DataPreviewActionTypes.stringTemplate) {
        let templateString = action.template;
        let templateFragments: string[] = [];

        // Handle template opener
        const openFragments = templateString.split("{{");
        for (const s of openFragments) {
            // Handle template closer
            if (s.includes("}}")) {
                const closeFragments = s.split("}}");
                closeFragments.forEach((sf) => {
                    if (sf !== "") {
                        templateFragments.push(sf);
                    }
                });
            } else {
                templateFragments.push(s);
            }
        }

        // TODO - annotate
        const finalFragments: string[] = [];
        templateFragments.forEach((tf) => {
            let finalFragment = "";
            Object.keys(data).forEach((key) => {
                if (tf.includes(`data.${key}`)) {
                    // TODO - handle non-string values here
                    finalFragment = String(data[key]);
                }
            });
            if (finalFragment) {
                finalFragments.push(finalFragment);
            } else {
                finalFragments.push(tf);
            }
        });

        // Join finalFragments and return
        return finalFragments.join("");
    }

    // Empty return statement
    return "";
}

/**
 * applyDataPreview
 * Applies DataPreview and returns resulting content
 * @param props.data - The data that's being inspected
 * @param props.dataPreview - The DataPreview being applied against props.data
 */
export function applyDataPreview(props: {
    data: Record<string, OptionValueInstance>;
    dataPreview: DataPreview;
}): string {
    const { data, dataPreview } = props;

    // Defines variable for return value
    let dataPreviewContent: null | string = null;

    // Iterates over each rule and evaluates accordingly
    dataPreview.rules.forEach((rule: DataPreviewRule) => {
        // Skip others if dataPreviewContent is already defined
        if (dataPreviewContent !== null) {
            return;
        }

        // Invoke applyDataPreviewRule, if constraint is met
        if (shouldApplyDataPreviewRule({ constraint: rule.constraint, data })) {
            dataPreviewContent = applyDataPreviewRule({
                action: rule.action,
                data,
            });
        }
    });

    // return content
    return dataPreviewContent || "";
}
