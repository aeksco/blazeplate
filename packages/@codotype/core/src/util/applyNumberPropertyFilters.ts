import { NumberPropertyFilters } from "../property-filter";

// // // //

/**
 * NumberPropertyFilterFunction
 * Type alais for an acceptable function for a NumberPropteryFilter
 */
type NumberPropertyFilterFunction = (num: number) => number;

/**
 * filterFunctions
 * Maps each value in NumberValueFilters enum to its respective function
 */
const filterFunctions: {
    [key in NumberPropertyFilters]: NumberPropertyFilterFunction;
} = {
    [NumberPropertyFilters.integerValue]: (num: number) => Math.floor(num),
    [NumberPropertyFilters.positiveValue]: (num: number) => Math.abs(num),
    [NumberPropertyFilters.negativeValue]: (num: number) => -1 * Math.abs(num),
};

/**
 * applyNumberPropertyFilters
 * Apply NumberValueFilters to a numeric value, in order of props.filters
 * @param props Options object
 * @param props.value The number value to apply the filters to
 * @param props.filters The array of NumberPropertyFilters to apply **in order**
 * @returns Filtered number value
 */
export function applyNumberPropertyFilters({
    value,
    filters,
}: {
    value: number;
    filters: NumberPropertyFilters[];
}): number {
    return filters.reduce(
        (str, filter) => filterFunctions[String(filter)](str),
        value,
    );
}
