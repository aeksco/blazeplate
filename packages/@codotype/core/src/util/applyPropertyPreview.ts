import {
    PropertyPreview,
    PropertyPreviewActionTypes,
    PropertyPreviewConstraintTypes,
    PropertyPreviewRule,
    PropertyPreviewConstraint,
    PropertyPreviewAction,
} from "../property-preview";
import { OptionValueInstance } from "../configuration-property";

// // // //

/**
 * shouldApplyPropertyPreviewRule
 * Returns a boolean indicating whether or not a specific PropertyPreviewRule should be applied against props.data
 * @param props.data - The data that's being evaluated by the PropertyPreviewRule
 * @param props.rule - The PropertyPreviewRule being evaluated
 */
export function shouldApplyPropertyPreviewRule(props: {
    data: Record<string, OptionValueInstance>;
    constraint: PropertyPreviewConstraint;
}): boolean {
    const { constraint, data } = props;

    const sourceValue: any = data[constraint.dataProperty];

    // Handle EQUALS
    if (constraint.type === PropertyPreviewConstraintTypes.equals) {
        if (sourceValue === constraint.value) {
            return true;
        }
    }

    // Handle contains
    if (
        constraint.type === PropertyPreviewConstraintTypes.contains &&
        typeof sourceValue === "string"
    ) {
        return sourceValue.includes(constraint.value);
    }

    // Handle exists
    if (constraint.type === PropertyPreviewConstraintTypes.exists) {
        if (sourceValue) {
            return true;
        }
    }

    // Return false by default
    return false;
}

/**
 * applyPropertyPreviewRule
 * @param props.data - The data that's being evaluated by the PropertyPreviewRule
 * @param props.rule - The PropertyPreviewRule being evaluated
 */
export function applyPropertyPreviewRule(props: {
    data: Record<string, OptionValueInstance>;
    action: PropertyPreviewAction;
}): string {
    const { action, data } = props;

    // Handle LITERAL
    if (action.type === PropertyPreviewActionTypes.literal) {
        return action.template;
    }

    // Handle STRING_TEMPLATE
    if (action.type === PropertyPreviewActionTypes.stringTemplate) {
        let templateString = action.template;
        let templateFragments: string[] = [];

        // Handle template opener
        const openFragments = templateString.split("{{");
        for (const s of openFragments) {
            // Handle template closer
            if (s.includes("}}")) {
                const closeFragments = s.split("}}");
                closeFragments.forEach((sf) => {
                    if (sf !== "") {
                        templateFragments.push(sf);
                    }
                });
            } else {
                templateFragments.push(s);
            }
        }

        // Iterate over templateFragments and perform replacement operations to produce finalFragments array
        const finalFragments: string[] = [];
        templateFragments.forEach((tf) => {
            let finalFragment = "";
            Object.keys(data).forEach((key) => {
                if (tf.includes(`data.${key}`)) {
                    finalFragment = String(data[key]);
                }
            });
            if (finalFragment) {
                finalFragments.push(finalFragment);
            } else {
                finalFragments.push(tf);
            }
        });

        // Join finalFragments and return
        return finalFragments.join("");
    }

    // Empty return statement
    return "";
}

/**
 * applyPropertyPreview
 * Applies PropertyPreview and returns resulting content
 * @param props.data - The data that's being inspected
 * @param props.propertyPreview - The PropertyPreview being applied against props.data
 */
export function applyPropertyPreview(props: {
    data: Record<string, OptionValueInstance>;
    propertyPreview: PropertyPreview;
}): string {
    const { data, propertyPreview } = props;

    // Defines variable for return value
    let propertyPreviewContent: null | string = null;

    // Iterates over each rule and evaluates accordingly
    propertyPreview.rules.forEach((rule: PropertyPreviewRule) => {
        // Skip others if propertyPreviewContent is already defined
        if (propertyPreviewContent !== null) {
            return;
        }

        // Invoke applyPropertyPreviewRule if constraint is met
        if (
            shouldApplyPropertyPreviewRule({
                constraint: rule.constraint,
                data,
            })
        ) {
            propertyPreviewContent = applyPropertyPreviewRule({
                action: rule.action,
                data,
            });
        }
    });

    // return content
    return propertyPreviewContent || "";
}
