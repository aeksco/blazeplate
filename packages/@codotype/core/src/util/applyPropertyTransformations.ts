import {
    StringPropertyTransformations,
    NumberPropertyTransformations,
} from "../property-transformation";
import { makeTitleCase } from "./makeTitleCase";
import { makeCamelCase } from "./makeCamelCase";
import { makeSnakeCase } from "./makeSnakeCase";
import { makePascalCase } from "./makePascalCase";
import { makeKebabCase } from "./makeKebabCase";

// // // //

/**
 * StringPropertyTransform
 * Type alais for an acceptable function for a StringPropteryFilter
 */
type StringPropertyTransform = (num: string) => string;

/**
 * stringTransforms
 * Maps each value in StringValueFilters enum to its respective function
 */
const stringTransforms: {
    [key in StringPropertyTransformations]: StringPropertyTransform;
} = {
    [StringPropertyTransformations.lowercase]: (val: string) =>
        val.toLowerCase(),
    [StringPropertyTransformations.uppercase]: (val: string) =>
        val.toUpperCase(),
    [StringPropertyTransformations.titlecase]: makeTitleCase,
    [StringPropertyTransformations.camelcase]: makeCamelCase,
    [StringPropertyTransformations.snakecase]: makeSnakeCase,
    [StringPropertyTransformations.pascalcase]: makePascalCase,
    [StringPropertyTransformations.kebabcase]: makeKebabCase,
    [StringPropertyTransformations.nonumbers]: (val: string) =>
        val.replace(/[0-9]/g, ""),
    [StringPropertyTransformations.nosymbols]: (val: string) =>
        val.replace(/[^a-zA-Z\s]/gi, ""),
    [StringPropertyTransformations.trimwhitespace]: (val: string) => val.trim(),
    [StringPropertyTransformations.removewhitespace]: (val: string) =>
        val.replace(/\s/g, ""),
};

/**
 * applyStringPropertyTransformations
 * Apply StringPropertyTransformations to a string value, in order of props.filters
 * @param props Options object
 * @param props.value The string value to apply the filters to
 * @param props.transformations The array of StringPropertyTransformations to apply **in order**
 * @returns Filtered string value
 */
export function applyStringPropertyTransformations({
    value,
    transformations,
}: {
    value: string;
    transformations: StringPropertyTransformations[];
}): string {
    return transformations.reduce(
        (str, filter) => stringTransforms[String(filter)](str),
        value,
    );
}

// // // //

/**
 * NumberPropertyTransform
 * Type alais for an acceptable function for a NumberPropteryFilter
 */
type NumberPropertyTransform = (num: number) => number;

/**
 * numberTransforms
 * Maps each value in NumberPropertyTransformations enum to its respective function
 */
const numberTransforms: {
    [key in NumberPropertyTransformations]: NumberPropertyTransform;
} = {
    [NumberPropertyTransformations.integerValue]: (num: number) =>
        Math.floor(num),
    [NumberPropertyTransformations.positiveValue]: (num: number) =>
        Math.abs(num),
    [NumberPropertyTransformations.negativeValue]: (num: number) =>
        -1 * Math.abs(num),
};

/**
 * applyNumberPropertyTransformations
 * Apply NumberPropertyTransformations to a numeric value, in order of props.filters
 * @param props Options object
 * @param props.value The number value to apply the filters to
 * @param props.transformations The array of NumberPropertyTransformations to apply **in order**
 * @returns Filtered number value
 */
export function applyNumberPropertyTransformations({
    value,
    transformations,
}: {
    value: number;
    transformations: NumberPropertyTransformations[];
}): number {
    return transformations.reduce(
        (str, filter) => numberTransforms[String(filter)](str),
        value,
    );
}
