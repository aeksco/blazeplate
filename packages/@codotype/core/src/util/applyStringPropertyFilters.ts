import { StringValueFilters } from "../property-filter";
import { makeTitleCase } from "./makeTitleCase";
import { makeCamelCase } from "./makeCamelCase";
import { makeSnakeCase } from "./makeSnakeCase";
import { makePascalCase } from "./makePascalCase";
import { makeKebabCase } from "./makeKebabCase";

/** Maps every StringValueFilter (hopefully) to its function. */
const filterFuncs = {
    [StringValueFilters.lowercase]: (val: string) => val.toLowerCase(),
    [StringValueFilters.uppercase]: (val: string) => val.toUpperCase(),
    [StringValueFilters.titlecase]: makeTitleCase,
    [StringValueFilters.camelcase]: makeCamelCase,
    [StringValueFilters.snakecase]: makeSnakeCase,
    [StringValueFilters.pascalcase]: makePascalCase,
    [StringValueFilters.kebabcase]: makeKebabCase,
    [StringValueFilters.nonumbers]: (val: string) => val.replace(/[0-9]/g, ""),
    [StringValueFilters.nosymbols]: (val: string) =>
        val.replace(/[^a-zA-Z\s]/gi, ""),
    [StringValueFilters.trimwhitespace]: (val: string) => val.trim(),
    [StringValueFilters.removewhitespace]: (val: string) =>
        val.replace(/\s/g, ""),
};

/**
 * applyStringPropertyFilters
 * Apply StringValueFilter(s) to a string value in order.
 * @param props Options object
 * @param props.value The string value to apply the filters to
 * @param props.filters The array of StringValueFilters to apply **in order**
 * @returns Filtered string value
 */
export function applyStringPropertyFilters({
    value,
    filters,
}: {
    value: string;
    filters: StringValueFilters[];
}): string {
    return filters.reduce(
        (str, filter) => filterFuncs[String(filter)](str),
        value,
    );
}
