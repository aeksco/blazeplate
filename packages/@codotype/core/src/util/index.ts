export * from "./indent";
export * from "./inflate";
export * from "./sanitizeTitle";
export * from "./trailingComma";
export * from "./validateGenerator";
export * from "./buildDefaultProject";
export * from "./makeCamelCase";
export * from "./makeSnakeCase";
export * from "./makeIdentifier";
export * from "./makeUniqueId";
export * from "./makeTitleCase";
export * from "./makeKebabCase";
export * from "./buildTokenCasing";
export * from "./buildTokenPluralization";
export * from "./schemaRenderer";
export * from "./applyPropertyTransformations";
export * from "./applyPropertyPreview";
