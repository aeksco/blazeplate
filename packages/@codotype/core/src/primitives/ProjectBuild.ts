import { ProjectBuild } from "../runtime";
import { ProjectInput } from "../project";

// // // //

interface ProjectBuildParams {
    id?: string;
    projectInput: ProjectInput;
}

export class ProjectBuildBuilder implements ProjectBuild {
    id: string = "";
    projectInput: ProjectInput;
    startTime: string = new Date(Date.now()).toUTCString();
    endTime: string = "";
    constructor(params: ProjectBuildParams) {
        this.id = params.id || this.id;
        this.projectInput = params.projectInput;
        this.startTime = this.startTime;
        this.endTime = this.endTime;
    }
}
