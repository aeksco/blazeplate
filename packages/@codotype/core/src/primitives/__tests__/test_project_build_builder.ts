import { ProjectBuildBuilder } from "../ProjectBuild";
import { ProjectInput } from "../../project";
// import { testState } from "../../";
import { buildDefaultProjectInput } from "../../util";
import { NextJsStarter } from "../../__tests__/test_state/plugins/NextJsStarter";

// // // //

describe("ProjectBuildBuilder", () => {
    test("works", () => {
        const projectInput: ProjectInput = buildDefaultProjectInput(
            NextJsStarter,
        );
        const projectBuild = new ProjectBuildBuilder({
            id: "",
            projectInput,
        });

        // Update projectBuild.startTime to prevent tests from failing
        // projectBuild.startTime = "1618439842359";

        expect(projectBuild).toMatchSnapshot();
    });
});
