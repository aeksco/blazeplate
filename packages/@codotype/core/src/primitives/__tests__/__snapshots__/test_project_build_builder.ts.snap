// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectBuildBuilder works 1`] = `
ProjectBuildBuilder {
  "endTime": "",
  "id": "",
  "projectInput": Object {
    "configuration": Object {
      "configure": Object {
        "analytics": Object {
          "googleAnalytics": false,
          "hotjar": false,
        },
        "hosting": Object {
          "platform": "vercel",
        },
        "seo": Object {
          "jsonLD": false,
          "opengraph": false,
          "sitemap": false,
          "twitter": false,
        },
        "tooling": Object {
          "eslint": true,
          "jest": false,
          "prettier": true,
          "storybook": false,
        },
      },
      "example-properties": Object {
        "boolean": false,
        "collection": Array [],
        "dropdown": "one",
        "instance": Object {
          "boolean": false,
          "text": "",
        },
        "number": "",
        "text": "",
      },
    },
    "id": "",
    "identifiers": Object {
      "camel": "projectName",
      "kebab": "project-name",
      "pascal": "ProjectName",
      "snake": "project_name",
      "title": "Project Name",
    },
    "pluginID": "react-next-ts-website-starter-variant",
    "pluginVersion": "0.0.1",
    "relations": Array [],
    "schemas": Array [],
  },
  "startTime": "Wed, 14 Apr 2021 22:43:24 GMT",
}
`;
